plugins/woocommerce/src/Internal/BatchProcessing/BatchProcessingController.php

```diff
@@ -0,0 +1,323 @@
+	 * @throws \Exception If processor class doesn't exist.
+	 */
+	private function get_processor_instance( string $processor_class_name ) : BatchProcessorInterface {
+		$processor = wc_get_container()->get( $processor_class_name );
+
+		/**
+		 * Filters the instance of processor for current class name.
+		 *
+		 * @since 6.7.0.
+		 */
+		$processor = apply_filters( 'woocommerce_get_batch_processor', $processor, $processor_class_name );
+		if ( ! isset( $processor ) && class_exists( $processor_class_name ) ) {
+			// This is a fallback for when the batch processor is not registered in the container.
+			$processor = new $processor_class_name();
+		}
+		if ( ! is_a( $processor, BatchProcessorInterface::class ) ) {
+			throw new \Exception( "Unable to initialize batch processor instance for $processor_class_name" );
+		}
+		return $processor;
+	}
+
+	/**
+	 * Helper method to get list of all pending processes.
+	 *
+	 * @return array List (of string) of pending processes classnames.
+	 */
+	public function get_pending() : array {
+		return get_option( self::PENDING_PROCESSES_OPTION_NAME, array() );
+	}
+
+	/**
+	 * Mark a pending process complete.
+	 *
+	 * @param string $processor Name of processor class which will be marked compete.
+	 */
+	private function mark_pending_process_complete( string $processor ) {
+		$pending_processes = $this->get_pending();
+		if ( in_array( $processor, $pending_processes, true ) ) {
+			$pending_processes = array_diff( $pending_processes, array( $processor ) );
+			$this->set_pending_processes( $pending_processes );
+		}
+	}
+
+	/**
+	 * Helper method set pending processes.
+	 *
+	 * @param array $processes List of classnames of all pending processes.
+	 */
+	private function set_pending_processes( array $processes ) {
+		update_option( self::PENDING_PROCESSES_OPTION_NAME, $processes, false );
+	}
+
+	/**
+	 * Check if a particular process is pending.
+	 *
+	 * @param string $processor_name Fully qualified class name of process.
+	 *
+	 * @return bool Whether the process is in progress.
+	 */
+	public function is_enqueued( string $processor_name ) : bool {
+		return in_array( $processor_name, $this->get_pending(), true );
+	}
+
+	/**
+	 * Stops all pending processes, and clears current memory state.
+	 */
+	public function force_clear_all_processes() {
+		as_unschedule_all_actions( self::SINGLE_BATCH_PROCESS_ACTION );
+		as_unschedule_action( self::CONTROLLER_CRON_NAME );
+		update_option( self::PENDING_PROCESSES_OPTION_NAME, array(), false );
+	}
+
+	/**
+	 * Log an error if happens during migration processing.
+	 *
+	 * @param \Exception              $error Exception object.
+	 * @param BatchProcessorInterface $batch_processor Batch processor instance.
+	 * @param array                   $batch Batch that finished processing.
+	 */
+	protected function log_error( \Exception $error, BatchProcessorInterface $batch_processor, array $batch ) : void {
+		$batch_detail_string = '';
```

Comment:
Suggestion for composing this string:

1. Check if all the batch items are numbers, and if so, use [ArrayUtil::to_ranges_string](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/src/Utilities/ArrayUtil.php#L104)
2. If not, check if first and last items are numbers or strings, and include them.
3. If not, revert to `print_r` or maybe even do not include nothing.

We can add some documentation recommending using numbers whenever possible, or strings if not, for the returned batches.

Reply:
I wonder if going through all data in batches and type checking will be a bit of a performance issue? I have added a filter on the final log message in https://github.com/woocommerce/woocommerce/pull/33233/commits/efb29e540a8e37f2507172d57d4f18582f55cf0b, maybe that's enough for this PR. wdyt?

Reply:
Keep in mind that this code is going to run only in case of error, so it shouldn't run often and it it runs it means that something is broken and thus performance won't really matter. But sure, a filter will do for now.